julia/src/api_types.jl

```diff
@@ -33,33 +33,38 @@ function bust_union(x::TaggedUnion{U}) where U
 end
 bust_union(x) = x  # not a union, leave it alone
 
+"Forces reinterpretting `raw` as a `T`. If sizeof(raw) <: size(T) high bits will be filled with junk."
 function force_reinterpret(::Type{T}, raw) where T
     isbits(raw) || throw(ArgumentError("Can only reinterpret from a isbits type"))
     isbitstype(T) || throw(ArgumentError("Can only reinterpret into a isbits type"))
     return unsafe_load(Ptr{T}(Base.pointer_from_objref(Ref(raw))))
 end
 
 
-struct CLit <: TaggedUnion{Tuple{Int64, Int32, Int8, Float64, Float32}}
+struct CLit <: TaggedUnion{Tuple{Int64, Int32, Int8, Float64, Float32, UInt32, UInt64}}
     tag::UInt64
     payload::NTuple{8, UInt8}  # actually the Union, needs to be big enough to hold largest which is an Int64
 end
+CLit(x) = CLit(CLit_tag(x), force_reinterpret(NTuple{8, UInt8}, x))
+
+CLit_tag(::Int64) = 0x0000_0000_0000_0000
+CLit_tag(::Int32) = 0x0000_0000_0000_0001
+CLit_tag(::Int8) = 0x0000_0000_0000_0002
+CLit_tag(::Float64) = 0x0000_0000_0000_0003
+CLit_tag(::Float32) = 0x0000_0000_0000_0004
+CLit_tag(::UInt32) = 0x0000_0000_0000_0005
+CLit_tag(::UInt64) = 0x0000_0000_0000_0006
```

Comment:
these could just be 0,1,2 etc.
The optimizer would do the conversions anyway.
But idk feels more honest to write them out in full as the types they are going to end up as.

Reply:
But they will have to get reinterpreted anyway, right? The tag is 1 byte wide, so this is way too many zeros. And it's not even binary?

Reply:
>the tag is 1 byte wide, so this is way too many zeros.

No? it is a `UInt64` it is 8 bytes long.
Isn't that what is going on here: https://github.com/google-research/dex-lang/blob/b80d02d7c063a9207175d5afd98f1f16fa984907/src/Dex/Foreign/Serialize.hs#L32-L33 ?

> And it's not even binary?

This is in hex as it uses the `0x` prefix (`0b` for binary).

Reply:
Ahhhh, right... Ok I must have mixed that up. Please ignore this comment.